# powerMRMA
Power package to examine the mediated path from gene to outcome through an intermediate phenotype. This R package compares MR and mediation analysis approaches to detect the path from the mediator to the outcome given at least one SNP serves as an instrumental variable for the mediator.

#### Installation
```
install.packages("devtools") # devtools must be installed first
install.packages("mediation")
install.packages("MendelianRandomization")

devtools::install_github("SharonLutz/powerMRMA")
```

#### Input

The methods to compare are designated with "methodnames", the default is methodnames = c("MR.Classical", "MR.Egger", "MR.IVW", "MR.Median", "MA.Imai", "MA.4Way").

The MR.Classical, MA.Imai and MA.4Way methods all test for the indirect effect from the first SNP input to Y through M. The MR.Egger, MR.IVW and MR.Median methods use all of the SNPs input to check for an indirect from the SNPs to Y through M.

The number of SNPs is set using "nSNP" (default = 4). X is generated from a binomial distribution where the minor allele frequency (MAF) for all SNPS is input as a vector, the default is MAF = c(0.2,0.2,0.2,0.2).

M is generated such that E\[M\] = \gamma0 + sum \gammaX * Xi + \gammaU

\gamma0 is set using "gamma0" (default = 0), \gammaX is set using "gammaX" and indicates the association of each SNP Xi with the mediator M. \gammaX must be input as a vector with length equal to nSNP, the default is \gammaX = c(0.15,0.15,0.15,0.15). The user can set \gamma_X to 0 for any SNPS that is not an instrumental variable of M. \gammaU is set with "gammaU" (default = 0), and indicates the level of unmeasured confounding to generate.

The variance of M is set with varM (default = 1). 

Y is generated such that E\[Y\] = ....

\beta0 is set with "beta0" (default = 0), \betaX is set with "betaX" and indicates the direct effect from each SNP Xi to the outcome Y. \betaX must be a vector of length equal to nSNP, the default is \betaX = c(0,0,0,0). \betaU is set with "betaU" (default = 0), and indicates the level of unmeasured confounding to generate. The association of M with Y is set using "betaM". \betaM must be a vector of length >= 2 and will be the x-axis of the power plot. 

The variance of Y is set with varY (default = 1).

U is generated such that U = rnorm(\muU,\varU) where the mean can be set with muU (default = 0) and the variance can be set with varU (default = 1).

An interaction between a SNP_i and M on Y can be generated by setting index i of "betaI" to be nonzero. \betaI must be a vector of length equal to nSNP, the default is no interaction \betaI = c(0,0,0,0).

Measurement error is generated such that M* = M + rnorm(\muME,\varME), and M* is used in analysis. \muME is set using "muME" (default = 0) and varME is set using "varME" (default = 1). Measurement error can be generated or not generated using "MeasurementError". TRUE indicates measurement error is generated while FALSE indicates no measurement error of the mediator is generated.

The plot name can be chosen by the user with "plot.name", the default is "powerMRMAplot". Additionally, the user can choose whether or not to include a legend with "legend.include", and can choose to have a color or grayscale plot with "legend.color".

The sample size is set using "n" (default = 1000). The number of simulations to run is set by "n.sim" (default = 500). Users should note that this function is slow.

The seed is set with "seed" (default = 1).

The default alpha level is 0.05 and can be changed by setting "alpha.level". The mediation analysis and MR Classical methods  are evaluated using an alpha level of alpha.level/nSNP while the MR Egger, MR IVW and MR Median methods are evaluated using alpha.level.

#### Example
This example displays the power of all six methods, using four SNPs as instrumental variables, each with a MAF of 0.2. Two levels of association between M and Y are evaluated, 0.15 and 0.25. There is no measurement error of the mediator or unmeasured confounding of the mediator outcome generated. There is no direct effect from any SNP X to the outcome Y or interaction between any X and M on Y generated. This code runs 100 simulations of a sample size of 1000.
```
library(powerMRMA)
?powerMRMA # For details on this function

powerMRMA (plot.name = "powerMRMAplot",methodnames = c("MR.Classical","MR.Egger","MR.IVW","MR.Median","MA.Imai","MA.4Way")
,n = 1000,n.sim=100,MeasurementError=F, nSNP = 4, MAF=c(0.2,0.2,0.2,0.2), gammaX = c(.15,.15,.15,.15), betaM = c(0, 0.2, 0.25))

```

#### Output
For this example analysis we get the following matrix of the power of each method to detect an indirect effect, and corresponding plot:

```
     MR.Classical MR.Egger MR.IVW MR.Median MA.Imai MA.4Way
[1,]         0.00     0.00   0.00      0.00     0.0     0.0
[2,]         0.01     0.05   0.26      0.06     0.5     0.4
[1] "The plot is saved in your home directory."
```
<img src="https://github.com/SharonLutz/powerMRMA/blob/master/powerMRMAplot.png" width="600">

#### Reference
Thwing A, Ghosh D, Hokanson JE, Lutz SM. (2018) Mediated Paths in Genetic Association Studies: A Comparison of Mendelian Randomization and Mediation Analysis Approaches. (Target Journal).

